File: Sources/AnagramEngine/Connections.swift
import Foundation

/// The web: every relation the app knows, radiating from one word.
///
/// Each tool in the app is one kind of edge between words — rearranged letters,
/// one-letter steps, shared sounds, phonetic overlaps, words hiding inside
/// words. `ConnectionWeb` gathers them all so a UI can show the whole
/// neighborhood at once and let you walk it: everything is connected.
public struct ConnectionWeb: Sendable {

    public enum Relation: String, CaseIterable, Sendable, Identifiable, Codable {
        case anagram = "Anagram"            // same letters, rearranged
        case oneLetter = "One letter away"  // change/add/drop a letter
        case homophone = "Homophone"        // sounds identical
        case rhyme = "Rhyme"                // shares the tail sound
        case fusion = "Fusion"              // overlaps by sound (brangel)
        case hidden = "Hidden inside"       // spelled inside the word
        case audible = "Heard inside"       // audible in the pronunciation
        public var id: String { rawValue }
    }

    public struct Node: Sendable, Identifiable, Hashable {
        public let word: String
        public let relation: Relation
        public let detail: String           // tooltip: why these are connected
        public var id: String { "\(relation.rawValue):\(word)" }

        public init(word: String, relation: Relation, detail: String) {
            self.word = word
            self.relation = relation
            self.detail = detail
        }
    }

    private let cryptic: CrypticHelper
    private let ladder: WordLadder
    private let phonetics: PhoneticDictionary?
    private let fusion: FusionFinder?

    public init(
        cryptic: CrypticHelper,
        ladder: WordLadder,
        phonetics: PhoneticDictionary?,
        fusion: FusionFinder?
    ) {
        self.cryptic = cryptic
        self.ladder = ladder
        self.phonetics = phonetics
        self.fusion = fusion
    }

    /// Up to `perRelation` neighbors of `word` for every relation type.
    public func connections(of word: String, perRelation: Int = 5) -> [Node] {
        let w = word.lowercased()
        var nodes: [Node] = []

        func add(_ words: some Sequence<String>, _ relation: Relation, _ detail: (String) -> String) {
            var seen = Set(nodes.map(\.word))
            seen.insert(w)
            var count = 0
            for other in words where !seen.contains(other) {
                guard count < perRelation else { break }
                nodes.append(Node(word: other, relation: relation, detail: detail(other)))
                seen.insert(other)
                count += 1
            }
        }

        add(cryptic.anagramWords(of: w).filter { $0 != w },
            .anagram) { "\($0): the letters of \(w), rearranged" }

        let steps = (ladder.changeOneLetter(w) + ladder.dropOneLetter(w) + ladder.addOneLetter(w))
        add(steps.sorted { ($0.count, $0) < ($1.count, $1) },
            .oneLetter) { "\($0): one letter away from \(w)" }

        if let phonetics {
            add(phonetics.homophones(w), .homophone) { "\($0): pronounced exactly like \(w)" }
            add(phonetics.rhymes(w).sorted { ($0.count, $0) < ($1.count, $1) },
                .rhyme) { "\($0): rhymes with \(w)" }
        }

        if let fusion {
            let fusions = fusion.fusions(of: w, minOverlap: 2, cap: perRelation * 2)
            var details: [String: String] = [:]
            for f in fusions where details[f.partner] == nil {
                details[f.partner] = "\(f.partner) ⋈ \(w) → “\(f.spelling)”"
            }
            add(fusions.map(\.partner), .fusion) { details[$0] ?? "sound-overlaps \(w)" }
            add(fusion.audibleWords(in: w), .audible) { "you can hear \($0) inside \(w)" }
        }

        add(cryptic.hiddenWords(in: w, minLength: 3).map(\.word).filter { $0 != w },
            .hidden) { "\($0) is spelled inside \(w)" }

        return nodes
    }
}

File: app/src/engine/connections.ts
// Ported from Sources/AnagramEngine/Connections.swift
import { CrypticHelper } from "./cryptic";
import { WordLadder } from "./wordLadder";
import { PhoneticDictionary } from "./phonetics";
import { FusionFinder } from "./fusion";
import { WordList } from "./wordList";

/** The nine kinds of connection between words. */
export type Relation =
  | "anagram"
  | "oneLetter"
  | "homophone"
  | "rhyme"
  | "fusion"
  | "hidden"
  | "audible"
  | "reversal"
  | "association";

export const ALL_RELATIONS: readonly Relation[] = [
  "anagram",
  "oneLetter",
  "homophone",
  "rhyme",
  "fusion",
  "hidden",
  "audible",
  "reversal",
  "association",
];

export interface ConnectionNode {
  word: string;
  relation: Relation;
  detail: string;
}

/** Semantic-neighbor provider, injected (data-driven on web, NLEmbedding in the Swift app). */
export type AssociationsProvider = (word: string, cap: number) => string[];

export interface ConnectionWebDeps {
  cryptic: CrypticHelper;
  ladder: WordLadder;
  phonetics?: PhoneticDictionary;
  fusion?: FusionFinder;
  /** needed for the reversal relation */
  words?: WordList;
  associations?: AssociationsProvider;
}

/**
 * Aggregates every relation type into one word-web: the neighbors of a word
 * across all nine dimensions at once, first-relation-wins dedup.
 */
export class ConnectionWeb {
  private readonly deps: ConnectionWebDeps;

  constructor(deps: ConnectionWebDeps) {
    this.deps = deps;
  }

  /**
   * Up to `perRelation` neighbors of `word` for every relation type.
   * `relations` restricts which dimensions are searched at all — a disabled
   * relation costs nothing and contributes nothing.
   */
  connections(
    word: string,
    perRelation = 5,
    relations: ReadonlySet<Relation> = new Set(ALL_RELATIONS),
  ): ConnectionNode[] {
    const w = word.toLowerCase();
    const { cryptic, ladder, phonetics, fusion, words, associations } = this.deps;
    const nodes: ConnectionNode[] = [];
    const seen = new Set([w]);

    const add = (candidates: string[], relation: Relation, detail: (c: string) => string): void => {
      let n = 0;
      for (const c of candidates) {
        if (n >= perRelation) break;
        if (seen.has(c)) continue;
        seen.add(c);
        nodes.push({ word: c, relation, detail: detail(c) });
        n++;
      }
    };

    if (relations.has("anagram")) {
      add(
        cryptic.anagramWords(w).filter((x) => x !== w),
        "anagram",
        (c) => `${c}: the letters of ${w}, rearranged`,
      );
    }

    if (relations.has("oneLetter")) {
      const steps = [...ladder.changeOneLetter(w), ...ladder.dropOneLetter(w), ...ladder.addOneLetter(w)];
      steps.sort((a, b) => (a.length !== b.length ? a.length - b.length : a < b ? -1 : 1));
      add(steps, "oneLetter", (c) => `${c}: one letter away from ${w}`);
    }

    if (phonetics) {
      if (relations.has("homophone")) {
        add(phonetics.homophones(w), "homophone", (c) => `${c}: pronounced exactly like ${w}`);
      }
      if (relations.has("rhyme")) {
        const rhymes = phonetics.rhymes(w);
        rhymes.sort((a, b) => (a.length !== b.length ? a.length - b.length : a < b ? -1 : 1));
        add(rhymes, "rhyme", (c) => `${c}: rhymes with ${w}`);
      }
    }

    if (fusion) {
      if (relations.has("fusion")) {
        const fusions = fusion.fusions(w, ["before", "after"], 2, perRelation * 2);
        const details = new Map<string, string>();
        for (const f of fusions) {
          if (!details.has(f.partner)) {
            details.set(f.partner, `${f.partner} ⋈ ${w} → “${f.spelling}”`);
          }
        }
        add(
          fusions.map((f) => f.partner),
          "fusion",
          (c) => details.get(c) ?? `sound-overlaps ${w}`,
        );
      }
      if (relations.has("audible")) {
        add(fusion.audibleWords(w), "audible", (c) => `you can hear ${c} inside ${w}`);
      }
    }

    if (relations.has("hidden")) {
      add(
        cryptic
          .hiddenWords(w, 3)
          .map((h) => h.word)
          .filter((x) => x !== w),
        "hidden",
        (c) => `${c} is spelled inside ${w}`,
      );
    }

    if (relations.has("reversal") && words) {
      const rev = [...w].reverse().join("");
      if (rev !== w && words.contains(rev)) {
        add([rev], "reversal", (c) => `${c}: ${w} spelled backwards`);
      }
    }

    if (relations.has("association") && associations) {
      add(associations(w, perRelation * 2), "association", (c) => `${c} keeps company with ${w}`);
    }

    return nodes;
  }
}
